File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/query/QuerySortBuilder.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazonaws.amplify.amplify_datastore.types.query

import com.amplifyframework.core.model.query.QuerySortBy
import com.amplifyframework.core.model.query.QuerySortOrder
import java.util.Locale

class QuerySortBuilder {
    companion object {
        @JvmStatic
        fun fromSerializedList(serializedList: List<Map<String, Any>>?): List<QuerySortBy>? {
            if (serializedList == null || serializedList.isEmpty()) {
                return null
            }
            return serializedList.map { fromSerializedMap(it) }
        }

        private fun fromSerializedMap(serializedMap: Map<String, Any>): QuerySortBy {
            return QuerySortBy(
                serializedMap["field"] as String,
                QuerySortOrder.valueOf(
                    (serializedMap["order"] as String).uppercase(Locale.ROOT)
                ) // android enums are upper case
            )
        }
    }
}

File: packages/amplify_datastore/ios/Classes/types/query/QuerySortBuilder.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import Foundation
import Amplify
import AmplifyPlugins

public enum QuerySortBuilder {
    static func fromSerializedList(_ serializedList: [[String: Any]]?) throws -> [QuerySortDescriptor]? {
        if let serializedList = serializedList, !serializedList.isEmpty {
            return try serializedList.map { try fromSerializedMap($0) }
        }
        return nil
    }

    static func fromSerializedMap(_ serializedMap: [String: Any]) throws -> QuerySortDescriptor {
        if let fieldName = serializedMap["field"] as? String, let order = serializedMap["order"] as? String {
            switch order {
            case "ascending":
                return QuerySortDescriptor(fieldName: fieldName, order: .ascending)
            case "descending":
                return QuerySortDescriptor(fieldName: fieldName, order: .descending)
            default:
                throw DataStoreError.decodingError("Failed to deserialize sort input from flutter",
                                                   "Please create an issue to amplify-flutter repo.")
            }
        }
        throw DataStoreError.decodingError("Failed to deserialize sort input from flutter",
                                           "Please create an issue to amplify-flutter repo.")
    }
}
